### PR TITLE
[FLINK-33817][flink-protobuf] Set ReadDefaultValues=False by default in proto3 for performance improvement

### PR DESCRIPTION
**Pull request by sharath1709**
## What is the purpose of the change

**Background**

 

The current Protobuf format [implementation](https://github.com/apache/flink/blob/c3e2d163a637dca5f49522721109161bd7ebb723/flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java) always sets ReadDefaultValues=True when using Proto3 version. This can cause severe performance degradation for large Protobuf schemas with OneOf fields as the entire generated code needs to be executed during deserialization even when certain fields are not present in the data to be deserialized and all the subsequent nested Fields can be skipped. Proto3 supports hasXXX() methods for checking field presence for non primitive types since Proto version [3.15](https://github.com/protocolbuffers/protobuf/releases/tag/v3.15.0). In the internal performance benchmarks in our company, we've seen almost 10x difference in performance for one of our real production usecase when allowing to set ReadDefaultValues=False with proto3 version. The exact difference in performance depends on the schema complexity and data payload but we should allow user to set readDefaultValue=False in general.

 

**Solution**

 
Set Default value for ReadDefaultValues=False for both proto3 and proto2 versions. Add documentation for users to set this value to true when using older protobuf versions before 3.15 and the performance implications. Additionally, We need to be careful to check for field presence only on non-primitive types if ReadDefaultValues is false and version used is Proto3. 

**NOTE:** This is a breaking change as older SQL queries without proper null checking can fail after this change. The row contents after deserialization will be different before and after this change. @libenchao will be making a note about this as part of the release notes.

## Brief change log

  - *[FLINK-33817][flink-protobuf] update protobuf docs about readDefaultValues*
  - *[FLINK-33817][flink-protobuf] Set ReadDefaultValues=False by default in Proto3 for performance improvement*

## Verifying this change

This change is already covered by existing unit tests and e2e tests for the Protobuf format for correctness. The performance improvement is benchmarked on testing clusters in my company.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? Both JavaDocs and flink docs are updated

### CODE REVIEW

**Comment by sharath1709** [comment]
@flinkbot run azure

**Comment by wckdman** [comment]
I'd like to upvote this PR since these changes also help in my use case: at the moment I can't distinguish one message from  another in `OneOf` field since there're a lot of default values. So the ability to filter fields with nulls is crucial.


I'm really looking forward to have this merged.

**Comment by sharath1709** [comment]
@wckdman Thanks for the upvote. I'm looking forward to getting this PR merged as well after getting the approval from the reviewers

**Comment by nathantalewis** [comment]
I just tested this on one of my clients' datasets which uses a lot of `oneof` fields and it fixed the same problem that wckdman mentioned.

**Comment by sharath1709** [comment]
@libenchao @maosuhan Gentle ping on the PR review

**Review by libenchao — COMMENTED** [review]
@sharath1709 the changes looks good in overall, just one comment about the document.

File: docs/content/docs/connectors/table/formats/protobuf.md
Comment (sharath1709) [reply docs/content/docs/connectors/table/formats/protobuf.md]: Ack, Updated the content to align with `PbFormatOptions`

Comment (sharath1709) [reply docs/content/docs/connectors/table/formats/protobuf.md]: @libenchao In case you haven't noticed, I'd also like to point out that this is a breaking change as older SQL queries without proper null checking can fail after this change. With that being said, this is a step in the right direction by making the default value to be the better performance choise

Comment (libenchao) [reply docs/content/docs/connectors/table/formats/protobuf.md]: @sharath1709 Thanks for the reminder, this is indeed a breaking change, we can add a release note to talk about it.

**Review by libenchao — APPROVED** [review]
LGTM

File: flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java
Comment (snuyanzin) [inline flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java:78]: since this class was touched nit:
the line below
```java
String uuid = UUID.randomUUID().toString().replaceAll("\\-", "");
```
could be replaced with

```java
String uuid = UUID.randomUUID().toString().replace("-", "");
```
to avoid compilation of patterns

**Comment by maosuhan** [comment]
@sharath1709 Thanks for the contribution.  I'll also help to look into this PR also and I expect to finish it this week. cc @libenchao

File: flink-formats/flink-protobuf/src/test/java/org/apache/flink/formats/protobuf/Pb3ToRowTest.java
Comment (sharath1709) [reply flink-formats/flink-protobuf/src/test/java/org/apache/flink/formats/protobuf/Pb3ToRowTest.java:112]: Nice catch, let me fix it

File: docs/content/docs/connectors/table/formats/protobuf.md
Comment (sharath1709) [reply docs/content/docs/connectors/table/formats/protobuf.md:152]: Yes, this is a breaking change and is called out in the comments. Let me update it in the PR description as well. @libenchao has mentioned that he would make a note about this in the release notes

File: flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java
Comment (sharath1709) [reply flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java:78]: Updated, thanks!

File: flink-formats/flink-protobuf/src/test/java/org/apache/flink/formats/protobuf/Pb3ToRowTest.java
Comment (sharath1709) [reply flink-formats/flink-protobuf/src/test/java/org/apache/flink/formats/protobuf/Pb3ToRowTest.java:112]: Done

**Comment by maosuhan** [comment]
LGTM now

**Comment by maosuhan** [comment]
> ## What is the purpose of the change
> **Background**
> 
> The current Protobuf format [implementation](https://github.com/apache/flink/blob/c3e2d163a637dca5f49522721109161bd7ebb723/flink-formats/flink-protobuf/src/main/java/org/apache/flink/formats/protobuf/deserialize/ProtoToRowConverter.java) always sets ReadDefaultValues=False when using Proto3 version. 

@sharath1709 I think the description should be "ProtoToRowConverter always sets ReadDefaultValues=true when using Proto3 version. "

**Comment by snuyanzin** [comment]
thanks for addressing comments, lgtm

**Comment by libenchao** [comment]
@maosuhan @snuyanzin Thanks for the review. Since you both have approved the PR, I'm merging now.

**Comment by ikstewa** [comment]
@sharath1709 I'm trying to follow up on this change to understand why we do not support optional primitive types. I believe proto3 has support for field presence on primitives when they have the `optional` keyword.

Do you have more details on _why_ we specifically do not support primitives?

> Additionally, We need to be careful to check for field presence only on non-primitive types if ReadDefaultValues is false and version used is Proto3.

**Pull request closed without merging**
